File: protocol/events.js
/**
 * Event Bus Client for Family VPN - JavaScript Implementation
 *
 * This is Layer 0 - the foundational event system for all UI components.
 * Mirror of protocol/events.go for use in Electron/browser contexts.
 *
 * Architecture:
 *
 *   ┌─────────────────────────────────────────────────────────────────┐
 *   │                      EventBus Client                            │
 *   ├─────────────────────────────────────────────────────────────────┤
 *   │  Features:                                                      │
 *   │    - Subscribe to namespaced events (e.g., "versions.*")       │
 *   │    - Receive snapshots for initial state                       │
 *   │    - Automatic reconnection with backoff                       │
 *   │    - Event replay from last sequence on reconnect              │
 *   └─────────────────────────────────────────────────────────────────┘
 *
 * Event Format:
 *
 *   {
 *     "ns": "versions.client_updated",
 *     "ts": 1234567890123,
 *     "seq": 42,
 *     "data": { ... }
 *   }
 */

// Namespace constants for type safety
const NS = {
  // System events
  SYSTEM_CONNECT: 'system.connect',
  SYSTEM_DISCONNECT: 'system.disconnect',
  SYSTEM_SNAPSHOT: 'system.snapshot',
  SYSTEM_ERROR: 'system.error',

  // Health events
  HEALTH_PING: 'health.ping',
  HEALTH_LATENCY: 'health.latency',
  HEALTH_UPTIME: 'health.uptime',

  // Version events
  VERSIONS_CLIENT: 'versions.client',
  VERSIONS_SERVER: 'versions.server',
  VERSIONS_SYNC: 'versions.sync',

  // Update events
  UPDATES_AVAILABLE: 'updates.available',
  UPDATES_APPLIED: 'updates.applied',
  UPDATES_REJECTED: 'updates.rejected',

  // Peer events
  PEERS_JOINED: 'peers.joined',
  PEERS_LEFT: 'peers.left',
  PEERS_UPDATED: 'peers.updated',

  // Dashboard events
  DASHBOARD_REFRESH: 'dashboard.refresh'
};

// Predefined subscription patterns
const SUB = {
  ALL: ['*'],
  VERSIONS: ['versions.*'],
  HEALTH: ['health.*'],
  PEERS: ['peers.*'],
  UPDATES: ['updates.*']
};

/**
 * Event represents a namespaced event from the server.
 */
class Event {
  constructor(data) {
    this.ns = data.ns;
    this.ts = data.ts;
    this.seq = data.seq;
    this.data = data.data;
    this.sid = data.sid || null;
  }

  /**
   * Check if this event matches a namespace pattern.
   */
  matches(pattern) {
    return matchPattern(pattern, this.ns);
  }

  /**
   * Get namespace parts (e.g., ['versions', 'client'] for 'versions.client')
   */
  parts() {
    return this.ns.split('.');
  }

  /**
   * Get namespace domain (first part)
   */
  domain() {
    return this.parts()[0];
  }

  /**
   * Get event type (second part)
   */
  type() {
    const p = this.parts();
    return p.length > 1 ? p[1] : null;
  }
}

/**
 * Snapshot represents consolidated state at a point in time.
 */
class Snapshot {
  constructor(data) {
    this.id = data.id;
    this.ts = data.ts;
    this.state = data.state || {};
    this.lastSeq = data.last_seq || 0;
  }

  /**
   * Get state for a specific namespace.
   */
  get(namespace) {
    return this.state[namespace];
  }

  /**
   * Check if snapshot has state for a namespace.
   */
  has(namespace) {
    return namespace in this.state;
  }
}

/**
 * EventBus client handles WebSocket connection and event subscriptions.
 */
class EventBusClient {
  constructor(options = {}) {
    this.url = options.url || null;
    this.subscriptions = options.subscriptions || ['*'];
    this.autoReconnect = options.autoReconnect !== false;
    this.reconnectDelay = options.reconnectDelay || 1000;
    this.maxReconnectDelay = options.maxReconnectDelay || 30000;

    this.ws = null;
    this.handlers = new Map(); // pattern -> [handlers]
    this.lastSeq = 0;
    this.currentSnapshot = null;
    this.reconnectAttempts = 0;
    this.connected = false;

    // Lifecycle callbacks
    this.onConnect = options.onConnect || (() => {});
    this.onDisconnect = options.onDisconnect || (() => {});
    this.onSnapshot = options.onSnapshot || (() => {});
    this.onError = options.onError || ((err) => console.error('[EventBus]', err));
  }

  /**
   * Connect to the WebSocket server.
   */
  connect(url) {
    if (url) this.url = url;
    if (!this.url) {
      throw new Error('EventBus: No WebSocket URL provided');
    }

    try {
      this.ws = new WebSocket(this.url);
      this._setupHandlers();
    } catch (err) {
      this.onError(err);
      this._scheduleReconnect();
    }
  }

  /**
   * Disconnect from the server.
   */
  disconnect() {
    this.autoReconnect = false;
    if (this.ws) {
      this.ws.close();
    }
  }

  /**
   * Subscribe to a namespace pattern.
   * @param {string} pattern - e.g., "versions.*", "health.ping"
   * @param {function} handler - Called with Event object
   * @returns {function} Unsubscribe function
   */
  subscribe(pattern, handler) {
    if (!this.handlers.has(pattern)) {
      this.handlers.set(pattern, []);
    }
    this.handlers.get(pattern).push(handler);

    // Send subscription to server if connected
    if (this.connected && !this.subscriptions.includes(pattern)) {
      this.subscriptions.push(pattern);
      this._sendSubscriptions();
    }

    // Return unsubscribe function
    return () => {
      const handlers = this.handlers.get(pattern);
      if (handlers) {
        const idx = handlers.indexOf(handler);
        if (idx >= 0) handlers.splice(idx, 1);
        if (handlers.length === 0) this.handlers.delete(pattern);
      }
    };
  }

  /**
   * Subscribe to multiple patterns at once.
   */
  subscribeMany(patterns, handler) {
    const unsubs = patterns.map(p => this.subscribe(p, handler));
    return () => unsubs.forEach(fn => fn());
  }

  /**
   * Get the current snapshot.
   */
  getSnapshot() {
    return this.currentSnapshot;
  }

  /**
   * Request a fresh snapshot from the server.
   */
  requestSnapshot() {
    if (this.connected) {
      this._send({ type: 'get_snapshot' });
    }
  }

  /**
   * Send a custom message to the server.
   */
  send(type, data = {}) {
    this._send({ type, ...data });
  }

  // ─────────────────────────────────────────────────────────────────
  // Private methods
  // ─────────────────────────────────────────────────────────────────

  _setupHandlers() {
    this.ws.onopen = () => {
      this.connected = true;
      this.reconnectAttempts = 0;
      console.log('[EventBus] Connected');

      // Send subscriptions and request initial state
      this._sendSubscriptions();
      this.onConnect();
    };

    this.ws.onclose = (event) => {
      this.connected = false;
      console.log('[EventBus] Disconnected', event.code, event.reason);
      this.onDisconnect(event);

      if (this.autoReconnect) {
        this._scheduleReconnect();
      }
    };

    this.ws.onerror = (error) => {
      console.error('[EventBus] WebSocket error:', error);
      this.onError(error);
    };

    this.ws.onmessage = (message) => {
      try {
        const data = JSON.parse(message.data);
        this._handleMessage(data);
      } catch (err) {
        console.error('[EventBus] Failed to parse message:', err);
      }
    };
  }

  _handleMessage(data) {
    // Handle different message types
    if (data.ns) {
      // This is an Event
      const event = new Event(data);
      this.lastSeq = Math.max(this.lastSeq, event.seq);
      this._dispatch(event);
    } else if (data.type === 'snapshot' || data.id) {
      // This is a Snapshot
      this.currentSnapshot = new Snapshot(data);
      this.lastSeq = this.currentSnapshot.lastSeq;
      this.onSnapshot(this.currentSnapshot);
    } else if (data.type === 'client_versions') {
      // Legacy compatibility: convert to event
      const event = new Event({
        ns: NS.VERSIONS_SYNC,
        ts: Date.now(),
        seq: ++this.lastSeq,
        data: {
          versions: data.versions,
          server_version: data.server_version
        }
      });
      this._dispatch(event);
    } else if (data.type === 'ping_update') {
      // Legacy compatibility: convert to event
      const event = new Event({
        ns: NS.HEALTH_PING,
        ts: data.timestamp || Date.now(),
        seq: ++this.lastSeq,
        data: {
          target: data.target,
          latency: data.latency
        }
      });
      this._dispatch(event);
    }
    // Add more legacy conversions as needed
  }

  _dispatch(event) {
    for (const [pattern, handlers] of this.handlers) {
      if (event.matches(pattern)) {
        handlers.forEach(h => {
          try {
            h(event);
          } catch (err) {
            console.error('[EventBus] Handler error:', err);
          }
        });
      }
    }
  }

  _send(data) {
    if (this.ws && this.ws.readyState === WebSocket.OPEN) {
      this.ws.send(JSON.stringify(data));
    }
  }

  _sendSubscriptions() {
    this._send({
      type: 'subscribe',
      namespaces: this.subscriptions,
      last_seq: this.lastSeq
    });
  }

  _scheduleReconnect() {
    this.reconnectAttempts++;
    const delay = Math.min(
      this.reconnectDelay * Math.pow(2, this.reconnectAttempts - 1),
      this.maxReconnectDelay
    );
    console.log(`[EventBus] Reconnecting in ${delay}ms (attempt ${this.reconnectAttempts})`);
    setTimeout(() => this.connect(), delay);
  }
}

/**
 * Check if a namespace matches a pattern.
 * Supports wildcards: "versions.*" matches "versions.client_updated"
 */
function matchPattern(pattern, namespace) {
  // Exact match
  if (pattern === namespace) return true;

  // Wildcard match
  if (pattern.endsWith('.*')) {
    const prefix = pattern.slice(0, -2);
    return namespace.startsWith(prefix + '.');
  }

  // Double wildcard matches everything
  if (pattern === '*' || pattern === '**') return true;

  return false;
}

/**
 * Create a new EventBus client.
 */
function createEventBus(options = {}) {
  return new EventBusClient(options);
}

// Export for both Node.js and browser
if (typeof module !== 'undefined' && module.exports) {
  module.exports = {
    NS, SUB, Event, Snapshot, EventBusClient, createEventBus, matchPattern
  };
} else if (typeof window !== 'undefined') {
  window.EventBus = {
    NS, SUB, Event, Snapshot, EventBusClient, createEventBus, matchPattern
  };
}

/*
 * Example Usage:
 *
 * // Create and connect
 * const bus = createEventBus({
 *   url: 'ws://10.8.0.1:9000/ws?vpn_ip=10.8.0.16',
 *   subscriptions: ['versions.*', 'health.*'],
 *   onSnapshot: (snapshot) => {
 *     console.log('Initial state:', snapshot.state);
 *     renderVersions(snapshot.get('versions'));
 *   }
 * });
 *
 * // Subscribe to specific events
 * bus.subscribe('versions.client', (event) => {
 *   console.log('Client version updated:', event.data);
 * });
 *
 * // Subscribe to all health events
 * bus.subscribe('health.*', (event) => {
 *   console.log('Health event:', event.ns, event.data);
 * });
 *
 * // Connect
 * bus.connect();
 *
 * // Later, disconnect
 * bus.disconnect();
 */

File: protocol/events.go
// Package protocol provides the event-driven architecture for Family VPN.
//
// This is Layer 0 - the foundational event bus that all components use.
// It provides namespaced events, subscriptions, and periodic snapshots.
//
// Architecture:
//
//	┌─────────────────────────────────────────────────────────────────┐
//	│                         Event Bus                               │
//	├─────────────────────────────────────────────────────────────────┤
//	│  Namespaces:                                                    │
//	│    system.*        - Core system events (connect, disconnect)   │
//	│    health.*        - Ping, latency, uptime events               │
//	│    versions.*      - Client/server version tracking             │
//	│    updates.*       - Hot-reload and update notifications        │
//	│    peers.*         - Peer discovery and status                  │
//	│    dashboard.*     - Dashboard-specific events                  │
//	├─────────────────────────────────────────────────────────────────┤
//	│  Snapshots:                                                     │
//	│    Every 5 minutes, consolidated state is broadcast             │
//	│    New clients receive snapshot immediately on connect          │
//	└─────────────────────────────────────────────────────────────────┘
//
// Event Format:
//
//	{
//	  "ns": "versions.client_updated",    // Namespace.EventType
//	  "ts": 1234567890123,                // Unix milliseconds
//	  "seq": 42,                          // Sequence number
//	  "data": { ... }                     // Event-specific payload
//	}
//
// Subscription Format (client -> server):
//
//	{
//	  "type": "subscribe",
//	  "namespaces": ["versions.*", "health.ping"]
//	}

package protocol

import (
	"encoding/json"
	"strings"
	"sync"
	"sync/atomic"
	"time"
)

// Event represents a namespaced event in the system.
// This is the canonical format - it will not change.
type Event struct {
	// Namespace is the event identifier (e.g., "versions.client_updated")
	Namespace string `json:"ns"`

	// Timestamp in Unix milliseconds
	Timestamp int64 `json:"ts"`

	// Sequence number for ordering (monotonically increasing)
	Sequence uint64 `json:"seq"`

	// Data is the event-specific payload
	Data interface{} `json:"data"`

	// SnapshotID links events to their snapshot epoch (optional)
	SnapshotID string `json:"sid,omitempty"`
}

// Snapshot represents a consolidated state at a point in time.
type Snapshot struct {
	// ID uniquely identifies this snapshot
	ID string `json:"id"`

	// Timestamp when snapshot was created
	Timestamp int64 `json:"ts"`

	// State is the consolidated state by namespace
	State map[string]interface{} `json:"state"`

	// LastSequence is the sequence number of the last event included
	LastSequence uint64 `json:"last_seq"`
}

// Subscription represents a client's event subscriptions.
type Subscription struct {
	// Patterns are namespace patterns (e.g., "versions.*", "health.ping")
	Patterns []string `json:"namespaces"`
}

// EventBus manages event publishing, subscriptions, and snapshots.
type EventBus struct {
	sequence    uint64
	subscribers map[string][]EventHandler
	subMutex    sync.RWMutex

	// Snapshot management
	currentSnapshot *Snapshot
	snapshotMutex   sync.RWMutex
	snapshotTicker  *time.Ticker
	snapshotDone    chan struct{}

	// State collectors for snapshot generation
	stateCollectors map[string]StateCollector
	collectorMutex  sync.RWMutex

	// Event history for replay (limited buffer)
	eventHistory []Event
	historyMutex sync.RWMutex
	maxHistory   int
}

// EventHandler is called when an event matches a subscription.
type EventHandler func(event Event)

// StateCollector provides current state for a namespace during snapshot.
type StateCollector func() interface{}

// NewEventBus creates a new event bus with snapshot interval.
func NewEventBus(snapshotInterval time.Duration) *EventBus {
	eb := &EventBus{
		subscribers:     make(map[string][]EventHandler),
		stateCollectors: make(map[string]StateCollector),
		eventHistory:    make([]Event, 0, 1000),
		maxHistory:      1000,
	}

	// Start snapshot ticker if interval > 0
	if snapshotInterval > 0 {
		eb.snapshotTicker = time.NewTicker(snapshotInterval)
		eb.snapshotDone = make(chan struct{})
		go eb.snapshotLoop()
	}

	return eb
}

// Stop shuts down the event bus.
func (eb *EventBus) Stop() {
	if eb.snapshotTicker != nil {
		eb.snapshotTicker.Stop()
		close(eb.snapshotDone)
	}
}

// snapshotLoop runs the periodic snapshot generator.
func (eb *EventBus) snapshotLoop() {
	for {
		select {
		case <-eb.snapshotTicker.C:
			eb.GenerateSnapshot()
		case <-eb.snapshotDone:
			return
		}
	}
}

// RegisterStateCollector registers a function to collect state for a namespace.
func (eb *EventBus) RegisterStateCollector(namespace string, collector StateCollector) {
	eb.collectorMutex.Lock()
	defer eb.collectorMutex.Unlock()
	eb.stateCollectors[namespace] = collector
}

// GenerateSnapshot creates a new snapshot from all registered collectors.
func (eb *EventBus) GenerateSnapshot() *Snapshot {
	eb.collectorMutex.RLock()
	state := make(map[string]interface{})
	for ns, collector := range eb.stateCollectors {
		state[ns] = collector()
	}
	eb.collectorMutex.RUnlock()

	snapshot := &Snapshot{
		ID:           time.Now().UTC().Format("20060102T150405Z"),
		Timestamp:    time.Now().UnixMilli(),
		State:        state,
		LastSequence: atomic.LoadUint64(&eb.sequence),
	}

	eb.snapshotMutex.Lock()
	eb.currentSnapshot = snapshot
	eb.snapshotMutex.Unlock()

	// Publish snapshot event
	eb.Publish("system.snapshot", snapshot)

	return snapshot
}

// GetSnapshot returns the current snapshot.
func (eb *EventBus) GetSnapshot() *Snapshot {
	eb.snapshotMutex.RLock()
	defer eb.snapshotMutex.RUnlock()
	return eb.currentSnapshot
}

// Subscribe registers a handler for events matching the pattern.
// Patterns support wildcards: "versions.*" matches all version events.
func (eb *EventBus) Subscribe(pattern string, handler EventHandler) {
	eb.subMutex.Lock()
	defer eb.subMutex.Unlock()
	eb.subscribers[pattern] = append(eb.subscribers[pattern], handler)
}

// Unsubscribe removes all handlers for a pattern.
func (eb *EventBus) Unsubscribe(pattern string) {
	eb.subMutex.Lock()
	defer eb.subMutex.Unlock()
	delete(eb.subscribers, pattern)
}

// Publish sends an event to all matching subscribers.
func (eb *EventBus) Publish(namespace string, data interface{}) Event {
	seq := atomic.AddUint64(&eb.sequence, 1)

	event := Event{
		Namespace: namespace,
		Timestamp: time.Now().UnixMilli(),
		Sequence:  seq,
		Data:      data,
	}

	// Add to history
	eb.historyMutex.Lock()
	if len(eb.eventHistory) >= eb.maxHistory {
		// Remove oldest 10%
		eb.eventHistory = eb.eventHistory[eb.maxHistory/10:]
	}
	eb.eventHistory = append(eb.eventHistory, event)
	eb.historyMutex.Unlock()

	// Dispatch to subscribers
	eb.dispatch(event)

	return event
}

// dispatch sends an event to all matching subscribers.
func (eb *EventBus) dispatch(event Event) {
	eb.subMutex.RLock()
	defer eb.subMutex.RUnlock()

	for pattern, handlers := range eb.subscribers {
		if matchPattern(pattern, event.Namespace) {
			for _, handler := range handlers {
				go handler(event) // Non-blocking dispatch
			}
		}
	}
}

// matchPattern checks if a namespace matches a pattern.
// Supports wildcards: "versions.*" matches "versions.client_updated"
func matchPattern(pattern, namespace string) bool {
	// Exact match
	if pattern == namespace {
		return true
	}

	// Wildcard match
	if strings.HasSuffix(pattern, ".*") {
		prefix := strings.TrimSuffix(pattern, ".*")
		return strings.HasPrefix(namespace, prefix+".")
	}

	// Double wildcard matches everything
	if pattern == "*" || pattern == "**" {
		return true
	}

	return false
}

// GetEventsSince returns events after the given sequence number.
func (eb *EventBus) GetEventsSince(seq uint64) []Event {
	eb.historyMutex.RLock()
	defer eb.historyMutex.RUnlock()

	var result []Event
	for _, e := range eb.eventHistory {
		if e.Sequence > seq {
			result = append(result, e)
		}
	}
	return result
}

// JSON serializes an event to JSON.
func (e Event) JSON() ([]byte, error) {
	return json.Marshal(e)
}

// MustJSON serializes or panics (for known-good events).
func (e Event) MustJSON() []byte {
	b, err := e.JSON()
	if err != nil {
		panic(err)
	}
	return b
}

// ParseEvent deserializes an event from JSON.
func ParseEvent(data []byte) (*Event, error) {
	var e Event
	if err := json.Unmarshal(data, &e); err != nil {
		return nil, err
	}
	return &e, nil
}

// ParseSubscription deserializes a subscription request.
func ParseSubscription(data []byte) (*Subscription, error) {
	var s Subscription
	if err := json.Unmarshal(data, &s); err != nil {
		return nil, err
	}
	return &s, nil
}

// MatchesAny checks if an event matches any of the subscription patterns.
func (s *Subscription) MatchesAny(namespace string) bool {
	for _, pattern := range s.Patterns {
		if matchPattern(pattern, namespace) {
			return true
		}
	}
	return false
}

// Common namespace constants for type safety.
const (
	// System events
	NSSystemConnect    = "system.connect"
	NSSystemDisconnect = "system.disconnect"
	NSSystemSnapshot   = "system.snapshot"
	NSSystemError      = "system.error"

	// Health events
	NSHealthPing    = "health.ping"
	NSHealthLatency = "health.latency"
	NSHealthUptime  = "health.uptime"

	// Version events
	NSVersionsClient = "versions.client"
	NSVersionsServer = "versions.server"
	NSVersionsSync   = "versions.sync"

	// Update events
	NSUpdatesAvailable = "updates.available"
	NSUpdatesApplied   = "updates.applied"
	NSUpdatesRejected  = "updates.rejected"

	// Peer events
	NSPeersJoined  = "peers.joined"
	NSPeersLeft    = "peers.left"
	NSPeersUpdated = "peers.updated"

	// Dashboard events
	NSDashboardRefresh = "dashboard.refresh"
)

// Predefined subscription patterns for common use cases.
var (
	SubAll      = Subscription{Patterns: []string{"*"}}
	SubVersions = Subscription{Patterns: []string{"versions.*"}}
	SubHealth   = Subscription{Patterns: []string{"health.*"}}
	SubPeers    = Subscription{Patterns: []string{"peers.*"}}
	SubUpdates  = Subscription{Patterns: []string{"updates.*"}}
)
